File: Approach/Sources/FeatureFlagsLibrary/FeatureFlag.swift
public struct FeatureFlag: Identifiable, Hashable {
	public let name: String
	public let introduced: String
	public let stage: RolloutStage
	public let isOverridable: Bool

	public var id: String { name }

	init(name: String, introduced: String, stage: RolloutStage, isOverridable: Bool = true) {
		self.name = name
		self.introduced = introduced
		self.stage = stage
		self.isOverridable = isOverridable
	}
}

extension FeatureFlag {
	public enum RolloutStage: Comparable, CaseIterable {
		case disabled
		case development
		case test
		case release

		public static func < (lhs: Self, rhs: Self) -> Bool {
			switch (lhs, rhs) {
			case (.disabled, .disabled): return false
			case (.disabled, _): return true
			case (.development, .disabled), (.development, .development): return false
			case (.development, _): return true
			case (.test, .disabled), (.test, .development), (.test, .test): return false
			case (.test, _): return true
			case (.release, _): return false
			}
		}
	}
}

File: android/core/featureflags/src/main/java/ca/josephroque/bowlingcompanion/core/featureflags/FeatureFlag.kt
package ca.josephroque.bowlingcompanion.core.featureflags

@Suppress("unused")
enum class RolloutStage {
	DISABLED,
	DEVELOPMENT,
	TEST,
	RELEASE,
}

interface Feature {
	val key: String
	val introduced: String
	val rolloutStage: RolloutStage
	val isOverridable: Boolean
}

enum class FeatureFlag(
	override val key: String,
	override val introduced: String,
	override val rolloutStage: RolloutStage,
	override val isOverridable: Boolean = true,
) : Feature {
	DEVELOPER_MODE("DevelopmentMode", "2024-08-23", RolloutStage.DEVELOPMENT, isOverridable = false),
	MANUAL_SERIES_FORM("ManualSeriesForm", "2024-03-28", RolloutStage.RELEASE),
	MANUAL_TEAM_SERIES_FORM("ManualTeamSeriesForm", "2024-08-30", RolloutStage.DEVELOPMENT),
	MOVING_SERIES_BETWEEN_LEAGUES("MovingSeriesBetweenLeagues", "2025-06-22", RolloutStage.RELEASE),
	PRE_BOWL_FORM("PreBowlForm", "2024-03-24", RolloutStage.RELEASE),
	REORDER_GAMES("ReorderGames", "2025-07-13", RolloutStage.RELEASE),
	SHARING_GAMES("SharingGames", "2024-11-03", RolloutStage.RELEASE),
	SHARING_SERIES("SharingSeries", "2024-11-03", RolloutStage.RELEASE),
}

internal fun FeatureFlag.isEnabled(): Boolean = if (BuildConfig.DEBUG) {
	this.rolloutStage >= RolloutStage.DEVELOPMENT
} else {
	this.rolloutStage == RolloutStage.RELEASE
}
